### PR TITLE
add tests for GORM scopes

### PR DESCRIPTION
**Pull request by saumil-oracle**
# Description

Add error handling and recovery tests
Add boundary condition tests (empty/nil scopes, conflicts)

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

**Test Configuration**:
* Firmware version:
* Hardware:
* Toolchain:
* SDK:

# Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by Harshit-Khasnis** [comment]
Please check the failed tests.

File: tests/scopes_test.go [hunk 1]
```diff
@@ -166,3 +169,376 @@ func TestComplexScopes(t *testing.T) {
...
+
+	// Test scope that tries to start its own transaction (nested transaction scenario)
+	nestedTxScope := func(db *gorm.DB) *gorm.DB {
+		return db.Begin()
+	}
+
+	err = DB.Transaction(func(tx *gorm.DB) error {
+		var users []User
+		return tx.Scopes(nestedTxScope).Find(&users).Error
+	})
```
[the remark this answers is not in the document]
Comment (saumil-oracle) [reply tests/scopes_test.go:415]: Added.

**Review by ting-lan-wang — APPROVED** [review]
